### PR TITLE
refactor deserialization to be safer

### PR DESCRIPTION
**Pull request by nextdude**
Several changes to deserialization to make it more robust

- `ConfluentAvroRegistryKafkaRecordDeserializationSchema` now respects the `specific.avro.reader` config. If you do not specify a setting, one will be selected based on whether the embedded avro record class is Generic or Specific. If you specify `specific.avro.reader = false` and the embedded avro record class is Specific, the schema will deserialize as a Generic record, then attempt to convert to a Specific record instance.
- The utility method that converts generic records to specific instances (`AvroUtils.toEmbeddedAvroInstance`) is now wrapped in a `Try`, as it may fail.
- Replace `AvroUtils.toSpecific` utility method implementation with call to `SpecificData.deepCopy()`.
- Processing of headers in the incoming kafka consumer record now takes account of the possibility of null header values (and ignores them) to avoid throwing a null pointer exception.
- `GlueAvroRegistryKafkaRecordDeserializationSchema` has a smarter key deserializer that will fallback to Confluent if you've configured a confluent fallback (so that string keys are properly processed).

### CODE REVIEW

File: src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala [hunk 1]
```diff
@@ -104,6 +104,7 @@ object AvroUtils extends LazyLogging {
           headers
         )
       )
+  }
 
   implicit class RichGenericRecord(genericRecord: GenericRecord) {
```
Comment (isterin) [inline src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala:109]: I think you can replace the whole toSpecific method with this:

` def toSpecific[A <: GenericRecord](instance: A): A = 
      SpecificData.get().deepCopy(instance.getSchema, genericRecord).asInstanceOf[A]`

Comment (isterin) [reply src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala:109]: I tested it briefly and it seems to work

Comment (isterin) [reply src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala:109]: Seems to work meaning, the way it's working now.  I do have an issue though with logical type conversions.  Still trying to figure it out.

Comment (nextdude) [reply src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala:109]: That's brilliant! Good find.

Comment (nextdude) [reply src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala:109]: Given that this can fail, we should wrap it in a `Try`, no?

Comment (isterin) [reply src/main/scala/io/epiphanous/flinkrunner/util/AvroUtils.scala:114]: Yes!

**Review by isterin — APPROVED** [review]
LGTM

**Pull request merged**